src/Manopt.jl

```diff
@@ -149,6 +149,25 @@ using Requires
 using SparseArrays
 using Statistics
 
+"""
+Rn(s::Symbol=:Manifolds, args; kwargs...)
+
+A small internal helper function to choose a Euclidean space.
+By default, this uses the [`DefaultManifold`](@extref ManifoldsBase.DefaultManifold) unless you load
+a more advanced Euclidean space like [`Euclidean`](@extref Manifolds.Euclidean)
+from [`Manifolds.jl`](@extref Manifolds.Manifolds)
+"""
+Rn(args...; kwargs...) = Rn(Val(Rn_default()), args...; kwargs...)
+@doc """
+   Rn_default()
+
+Specify a default value to dispatch [`Rn`](@ref) on.
+This default is set to `Manifolds`, indicating, that when this package is loded,
+it is the preferred package to ask for a Euclidean space.
+"""
+Rn_default() = :Manifolds
+Rn(::Val{T}, args...; kwargs...) where {T} = DefaultManifold(args...; kwargs...)
```

Comment:
Wouldn't it be more simple to just have `Euclidean` in ManifoldsBase.jl? Or just use `DefaultManifold` regardless of whether Manifolds.jl is loaded?

Reply:
Well we discussed the Euclidean part lengthy recently ;)
Another argument against moving `Euclidean` is, that it would split the implementation into two (the basic functions in base the newer / developing ones in Manifolds).

I would keep this function internal here, and it allows mainly a nicer display (Euclidean) when manifolds is loaded, while the algorithm still works if it its not loaded.
Actually this is the first algorithm that does need these product manifolds with Euclidean (3 times even).

Why do you not like this idea?

Reply:
We did, and this right here is IMO the perfect example why `Euclidean` in ManifoldsBase would be useful -- something the recent discussion was somewhat lacking.

Anyway, if you think this is the right solution, maybe we should have this function in ManifoldsBase and exported? And if you don't think this is the right solution, then you have your question answered.

Reply:
I think this it the right solution.

We could call it differently, when we move this to Base (maybe even `Euclidean` – ah that might not work when it should be overridden by Manifolds – maybe `vector_space(n)`?).

Reply:
> I'm not necessarily for having Euclidean in ManifoldsBase but it would be nice to have an official solution for this exact problem.

You are right a function like `vector_space(n)` is probably a very good tradeoff, that allows to access Euclidean (or DefaultManifold) from ManifoldsBase.
The more I see that, the more I like `vector_space(n)` for this.

Reply:
`vector_space` sounds like a good name (definitely better than `Rn`) but for consistency I think we can't have it return `DefaultManifold` from ManifoldsBase.jl by default. We can only tell users that if they need to call it without loading `Manifolds.jl`, they have to implement it themselves (possibly as returning `DefaultManifold`).

Reply:
Can you elaborate as to why? I like in my solution here, that it does exactly that.
It always provides a vector space, `DefaultManifold` if nothing more precise given, `Euclidean` if available. That's what the `Rn_default` is for.

Reply:
Mostly because it makes the value returned by `vector_space(n)` dependent on module load order, which can lead to surprising results where seemingly trivial changes change the type you get. It may have really weird side effects if for some reason IPM is used both before and after loading Manifolds.jl.

IMO the minor convenience is not worth trouble it can cause.

Reply:
Either `DefaultManifold` is a good enough result of `vector_space` and there is not point in changing it to `Euclidean`, or it is not and should not be returned anyway.

Reply:
Hm, but then there is no need to introduce `vector_spacee(n)` basically? If `DefaultManifold` is a good enough default we Cana also just write that anywhere. 
I functionality I even agree, it is. The main reason I did this here is that `Euclidean` looks nicer in prints.

You are right that there is a changing behaviour when loading a package – that might be misleading / confusing to a user. I would be fine with the convenience over the trouble, but I totally see that is a matter of taste.

We could turn the `Rn_default` into a `ManifoldsBase.DEFAULT_VECTORSPACE` setting or so? With my persona preference I could still set that to `:Euclidean` (or `:Manifolds`, depending on whether we consider that a manifold or package indicator). Though I am not sure a variable or a function is the best choice for that, this would allow to default it to `:ManifoldsBase`(or `:DefaultManifold`) to keep your preference by default.
On the other hand the bad-kellertuer-preference in Manopt.j could be to change that to `:Manifolds` ( or `:Euclidean`) for my convenience (and the users confusion 🤓)

Reply:
> Hm, but then there is no need to introduce `vector_spacee(n)` basically? If `DefaultManifold` is a good enough default we Cana also just write that anywhere.
> I functionality I even agree, it is. The main reason I did this here is that `Euclidean` looks nicer in prints.

Yes, if `vector_space` returns `DefaultManifold`, then there is not much reason for it to exist. IMO it only makes sense if it's left unimplemented. And `Euclidean` looking nicer is a fairly weak reason IMO to introduce such surprising behavior.

> We could turn the `Rn_default` into a `ManifoldsBase.DEFAULT_VECTORSPACE` setting or so? With my persona preference I could still set that to `:Euclidean` (or `:Manifolds`, depending on whether we consider that a manifold or package indicator). Though I am not sure a variable or a function is the best choice for that, this would allow to default it to `:ManifoldsBase`(or `:DefaultManifold`) to keep your preference by default. On the other hand the bad-kellertuer-preference in Manopt.j could be to change that to `:Manifolds` ( or `:Euclidean`) for my convenience (and the users confusion 🤓)

That isn't much better because each time such global setting is used, type instability would be necessary. We do a similar thing for AD backends but that's a different story because we tell the users that it's only for convenience and for performance they should be explicit, and there are actual substantial differences between backends.

If you really want to see `Euclidean` instead of `DefaultManifold`, maybe just implement IPM in `ManoptManifoldsExt`? Then it wouldn't have to touch `DefaultManifold` at all. You've even said recently that `DefaultManifold` should only be used for testing. So it would make its use in IPM not valid, and so IPM should only be available with Manifolds.jl loaded. This would also be a valid solution.

Reply:
Since this would also only happen on initialisation, I do not see much of a performance issue here. All manifolds (product manifolds) I need are stored.

So with that, something like `vector_space(n)` (even in ManifoldsBase) for people to specify their favourite vector space implementation is something I think is nice (however realised)

Reply:
I can see a few different valid designs, just the one where we use either `DefaultManifold` or `Euclidean` based on whether `Manifolds.jl` is loaded or not is one that makes the least sense to me. These are IMO valid solutions:
1. IPM uses `vector_space` function, which is unimplemented in `ManifoldsBase.jl` and returns `Euclidean` when Manifolds.jl is loaded.
2. IPM has `vector_space` as an input argument.
3. IPM uses `DefaultManifold` explicitly.
4. IPM uses `Euclidean` explicitly and thus requires `Manifolds.jl`.

> To be honest IPM could be available as soon as anyone gives me a vector space (in addition to an arbitrary manifold) so I do not agree to put it into the extension.

Then `vector_space` should be an input argument of IPM, not a global function. It could have a default value based on whether Manifolds.jl is loaded or not, which would be somewhat acceptable because it would be limited in scope to IPM.

Reply:
I think I like 2, but would really prefer what you wrote in your last part, that this default value is improved when loading Manifolds.jl (sure only in display actually). I am fine if that is only for IPM and hence limited scope, as it is now with `Rn`. So the main point here is, do we want to offer such thing in `ManifoldsBase`?

Reply:
OK, then I think the conclusion is clear here and we should go with 2 but it means that `vector_space` isn't going to be in ManifoldsBase, and every algorithm that requires some kind of Euclidean space should have their own separate implementation of option 2. The last part is basically a small extension of option 2.

Reply:
Let's close this since we have the solution remark below open still.

Reply:
I think my comment below should illustrate the solution well enough :slightly_smiling_face: . I wouldn't call your idea "not possible", it's just more confusing than it needs to be with no way out. The `vector_space` argument gives a way out so it makes the design good enough.

Reply:
As I said, I do not fully understand the difference, but I am fully fine with the `vector_space` argument idea and will follow that. Maybe I was not clear enough above, but the vector_space = thing makes my idea possible as long as I can keep my `Rn`.